# Machine Learning Zoomcamp
## Homework
https://github.com/DataTalksClub/machine-learning-zoomcamp/blob/master/cohorts/2024/01-intro/homework.md

In [4]:
import pandas as pd
import numpy as np
from numpy.linalg import inv, pinv

## Pandas Version
Q1. Pandas version
What's the version of Pandas that you installed?

You can get the version information using the __version__ field:

In [3]:
pd.__version__

'2.2.2'

# Read csv

In [5]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/laptops.csv')

In [3]:
display(df)

Laptop       Status   Brand  \
0     ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...          New    Asus   
1     Alurin Go Start Intel Celeron N4020/8GB/256GB ...          New  Alurin   
2     ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...          New    Asus   
3     MSI Katana GF66 12UC-082XES Intel Core i7-1270...          New     MSI   
4     HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...          New      HP   
...                                                 ...          ...     ...   
2155  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2156  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2157  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2158  Razer Book 13 Intel Evo Core i7-1165G7/16GB/1T...  Refurbished   Razer   
2159  Razer Book FHD+ Intel Evo Core i7-1165G7/16GB/...  Refurbished   Razer   

           Model                CPU  RAM  Storage Storage type       GPU  \
0     ExpertBook      Intel Core i5    8      512          SSD       NaN   
1             Go      Intel Celeron    8      256          SSD       NaN   
2     ExpertBook      Intel Core i3    8      256          SSD       NaN   
3         Katana      Intel Core i7   16     1000          SSD  RTX 3050   
4            15S      Intel Core i5   16      512          SSD       NaN   
...          ...                ...  ...      ...          ...       ...   
2155       Blade      Intel Core i7   16     1000          SSD  RTX 3060   
2156       Blade      Intel Core i7   16     1000          SSD  RTX 3070   
2157       Blade      Intel Core i7   32     1000          SSD  RTX 3080   
2158        Book  Intel Evo Core i7   16     1000          SSD       NaN   
2159        Book  Intel Evo Core i7   16      256          SSD       NaN   

      Screen Touch  Final Price  
0       15.6    No      1009.00  
1       15.6    No       299.00  
2       15.6    No       789.00  
3       15.6    No      1199.00  
4       15.6    No       669.01  
...      ...   ...          ...  
2155    17.3    No      2699.99  
2156    17.3    No      2899.99  
2157    17.3    No      3399.99  
2158    13.4   Yes      1899.99  
2159    13.4   Yes      1699.99  

[2160 rows x 12 columns]

# Records Count
Q2. Records count
How many records are in the dataset?

- [ ] 12
- [ ] 1000
- [X] 2160
- [ ] 12160

In [4]:
df.shape

(2160, 12)

Q3. Laptop brands
How many laptop brands are presented in the dataset?

- [ ] 12
- [ ] 27
- [ ] 28
- [x] 2160

In [9]:
len(df['Brand'].unique())

27

Q4. Missing values
How many columns in the dataset have missing values?

- [ ] 0
- [x] 1
- [ ] 2
- [ ] 3

In [5]:
df.count()

Laptop          2160
Status          2160
Brand           2160
Model           2160
CPU             2160
RAM             2160
Storage         2160
Storage type    2118
GPU              789
Screen          2156
Touch           2160
Final Price     2160
dtype: int64

In [10]:
df.isnull().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

In [11]:
df.isna().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

# Maximum final price
Q5. Maximum final price
What's the maximum final price of Dell notebooks in the dataset?

- [ ] 869
- [ ] 3691
- [ ] 3849
- [x] 3936

In [6]:
df_dell=df[df['Brand']=="Dell"]
print(df_dell['Final Price'].max())

3936.0


# Median value of Screen
Q6. Median value of Screen
Find the median value of Screen column in the dataset.
Next, calculate the most frequent value of the same Screen column.
Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.
Now, calculate the median value of Screen once again.
Has it changed?

Hint: refer to existing mode and median functions to complete the task.

- [ ] Yes
- [x] No

In [31]:
print(df['Screen'].median())

15.6


In [7]:
print(df['Screen'].value_counts())

Screen
15.60    1009
14.00     392
16.00     174
17.30     161
13.30     131
16.10      48
17.00      33
13.00      27
15.00      21
13.40      19
13.50      19
11.60      16
14.20      14
12.30      13
14.10      11
13.60      11
16.20      10
15.30       8
10.50       7
12.40       6
14.40       6
15.40       5
12.00       4
18.00       3
14.50       3
13.90       2
12.50       1
10.95       1
10.10       1
Name: count, dtype: int64


In [8]:
print(df['Screen'].mode())

0    15.6
Name: Screen, dtype: float64


In [9]:
df['Screen']=df['Screen'].fillna(15.6)
print(df['Screen'].mode())

0    15.6
Name: Screen, dtype: float64


# Sum of weights
Q7. Sum of weights
1. Select all the "Innjoo" laptops from the dataset.
1. Select only columns RAM, Storage, Screen.
1. Get the underlying NumPy array. Let's call it X.
1. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
1. Compute the inverse of XTX.
1. Create an array y with values [1100, 1300, 800, 900, 1000, 1100].
1. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
1. What's the sum of all the elements of the result?
Note: You just implemented linear regression. We'll talk about it in the next lesson.

- [ ] 0.43
- [ ] 45.29
- [ ] 45.58
- [x] 91.30

In [11]:
df_innjoo = df[df['Brand']=='Innjoo']

In [12]:
display(df_innjoo)

Laptop Status   Brand Model  \
1478  InnJoo Voom Excellence Intel Celeron N4020/8GB...    New  Innjoo  Voom   
1479  InnJoo Voom Excellence Pro Intel Celeron N4020...    New  Innjoo  Voom   
1480  Innjoo Voom Intel Celeron N3350/4GB/64GB eMMC/...    New  Innjoo  Voom   
1481  Innjoo Voom Laptop Max Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1482  Innjoo Voom Laptop Pro Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1483  Innjoo Voom Pro Intel Celeron N3350/6GB/128GB ...    New  Innjoo  Voom   

                CPU  RAM  Storage Storage type  GPU  Screen Touch  Final Price  
1478  Intel Celeron    8      256          SSD  NaN    15.6    No       311.37  
1479  Intel Celeron    8      512          SSD  NaN    15.6    No       392.55  
1480  Intel Celeron    4       64         eMMC  NaN    14.1    No       251.40  
1481  Intel Celeron    6       64         eMMC  NaN    14.1    No       383.61  
1482  Intel Celeron    6      128          SSD  NaN    14.1    No       317.02  
1483  Intel Celeron    6      128         eMMC  NaN    14.1    No       431.38

In [14]:
df_injoo=df_innjoo[['RAM','Storage', 'Screen']]

In [15]:
display(df_injoo)

RAM  Storage  Screen
1478    8      256    15.6
1479    8      512    15.6
1480    4       64    14.1
1481    6       64    14.1
1482    6      128    14.1
1483    6      128    14.1

In [26]:
X=df_injoo.to_numpy()

In [27]:
print(X.T)

[[  8.    8.    4.    6.    6.    6. ]
 [256.  512.   64.   64.  128.  128. ]
 [ 15.6  15.6  14.1  14.1  14.1  14.1]]


#### X multiplied by transpose X

In [29]:
XTX=np.matmul(X.T, X)

In [30]:
print(XTX)

[[2.52000e+02 8.32000e+03 5.59800e+02]
 [8.32000e+03 3.68640e+05 1.73952e+04]
 [5.59800e+02 1.73952e+04 1.28196e+03]]


#### Inverse of X multiplied by transpose X

In [20]:
iXTX=inv(XTX)

In [21]:
y=[1100, 1300, 800, 900, 1000, 1100]

In [22]:
w=np.matmul(np.matmul(iXTX, X.T),y)

In [24]:
w.sum()

np.float64(91.2998806299555)